File: LLD/RideSharingSystem.java
package com.uber.lld;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/*
 * ==============================================================================================
 * RIDE SHARING SERVICE (LIKE UBER) LOW LEVEL DESIGN
 * ==============================================================================================
 * 
 * Key Features:
 * 1. User Management: Rider and Driver (Extends User).
 * 2. Geo-Location: Tracking driver locations.
 * 3. Matching Service: Assign closest AVERAGE driver to Rider.
 * 4. Trip Lifecycle: REQUESTED -> ASSIGNED -> ON_TRIP -> COMPLETED.
 * 5. Cost Calculation: Strategy for different ride types (Regular/Premium).
 * 
 * Design Patterns:
 * 1. Observer: Rider gets notified when Driver status changes.
 * 2. Strategy: PricingStrategy (Surge/RideType).
 * 3. Singleton: UberSystem (Facade).
 * 4. State: TripStatus.
 *
 * Class Design Diagram:
 * ---------------------
 * [UberSystem] "1" *-- "*" [Rider]
 * [UberSystem] "1" *-- "*" [Driver]
 * [UberSystem] "1" *-- "1" [TripManager]
 * [TripManager] "1" *-- "*" [Trip]
 * [Trip] "1" *-- "1" [Rider]
 * [Trip] "1" *-- "1" [Driver]
 * [Trip] "1" *-- "1" [PricingStrategy]
 * [Driver] ..> [Location]
 * [Driver] ..> [DriverStatus]
 * [Trip] ..> [TripStatus]
 *
 * Class Details:
 * ---------------------
 * 1. UberSystem (Facade)
 *    - Role: Main entry point.
 *    - Methods: requestRide(), completeTrip().
 *
 * 2. TripManager
 *    - Role: Handles matching logic.
 *    - Methods: findDriver(), createTrip().
 *
 * 3. Driver
 *    - Attributes: currentLocation, status (AVAILABLE/BUSY).
 *
 * 4. Location
 *    - Attributes: lat, lon.
 *    - Methods: distanceTo().
 */

public class RideSharingSystem {
    public static void main(String[] args) {
        System.out.println("--- Uber System Demo ---");
        
        UberSystem system = UberSystem.getInstance();

        // 1. Register Users
        Rider rider = new Rider("R1", "Alice", new Location(0, 0)); // At Origin
        Driver d1 = new Driver("D1", "Bob", new Location(1, 1)); // Close
        Driver d2 = new Driver("D2", "Charlie", new Location(10, 10)); // Far
        
        system.addDriver(d1);
        system.addDriver(d2);

        // 2. Request Ride
        System.out.println("\n[Action] Alice requests ride from (0,0) to (5,5)");
        Location dropOff = new Location(5, 5);
        
        try {
            Trip trip = system.requestRide(rider, rider.getLocation(), dropOff, RideType.REGULAR);
            System.out.println("Trip Created: " + trip.getId());
            System.out.println("Driver Assigned: " + trip.getDriver().getName());
            
            // 3. Simulate Trip
            System.out.println("\n[Action] Driver starts trip...");
            trip.setStatus(TripStatus.ON_TRIP);
            
            System.out.println("[Action] Trip Completed.");
            system.completeTrip(trip.getId());
            
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

// ==========================================
// Enums & Strategy
// ==========================================

enum RideType { REGULAR, PREMIUM }
enum TripStatus { REQUESTED, ASSIGNED, ON_TRIP, COMPLETED, CANCELLED }
enum DriverStatus { AVAILABLE, ON_TRIP, OFFLINE }

interface PricingStrategy {
    double calculateFare(double distance, double time);
}

class RegularPricing implements PricingStrategy {
    @Override
    public double calculateFare(double distance, double time) {
        return (distance * 10) + (time * 1); // Base calculation
    }
}

class PremiumPricing implements PricingStrategy {
    @Override
    public double calculateFare(double distance, double time) {
        return (distance * 20) + (time * 2); // 2x cost
    }
}

// ==========================================
// Domain Models
// ==========================================

class Location {
    double x, y;
    public Location(double x, double y) { this.x = x; this.y = y; }
    
    public double distanceTo(Location other) {
        return Math.sqrt(Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2));
    }
    
    @Override public String toString() { return String.format("(%.1f, %.1f)", x, y); }
}

abstract class User {
    String id;
    String name;
    public User(String id, String name) { this.id = id; this.name = name; }
    public String getName() { return name; }
}

class Rider extends User {
    Location currentLocation;
    public Rider(String id, String name, Location loc) { 
        super(id, name); 
        this.currentLocation = loc;
    }
    public Location getLocation() { return currentLocation; }
}

class Driver extends User {
    Location currentLocation;
    DriverStatus status;
    double rating; // Extensibility

    public Driver(String id, String name, Location loc) {
        super(id, name);
        this.currentLocation = loc;
        this.status = DriverStatus.AVAILABLE;
        this.rating = 5.0;
    }
    
    public void setStatus(DriverStatus s) { this.status = s; }
    public void setLocation(Location l) { this.currentLocation = l; }
    public Location getLocation() { return currentLocation; }
    public DriverStatus getStatus() { return status; }
}

class Trip {
    String id;
    Rider rider;
    Driver driver;
    Location src;
    Location dest;
    TripStatus status;
    PricingStrategy pricing;
    double fare;

    public Trip(String id, Rider r, Driver d, Location s, Location dst, PricingStrategy p) {
        this.id = id; this.rider = r; this.driver = d; 
        this.src = s; this.dest = dst; this.pricing = p;
        this.status = TripStatus.ASSIGNED;
    }
    
    public void setStatus(TripStatus s) { 
        this.status = s; 
        System.out.println("Trip " + id + " status updated to: " + s);
        if(s == TripStatus.COMPLETED) {
            calculateFare();
        }
    }
    
    private void calculateFare() {
        double dist = src.distanceTo(dest);
        this.fare = pricing.calculateFare(dist, 10); // Assume 10 mins for demo
        System.out.println("Total Fare: $" + String.format("%.2f", fare));
    }
    
    public String getId() { return id; }
    public Driver getDriver() { return driver; }
}

// ==========================================
// System Manager
// ==========================================

class TripManager {
    public Driver findNearestDriver(Location pickup, List<Driver> drivers) {
        Driver best = null;
        double minDistance = Double.MAX_VALUE;
        
        for(Driver d : drivers) {
            if(d.getStatus() == DriverStatus.AVAILABLE) {
                double dist = d.getLocation().distanceTo(pickup);
                if(dist < minDistance) {
                    minDistance = dist;
                    best = d;
                }
            }
        }
        return best;
    }
}

class UberSystem {
    private static UberSystem instance;
    private List<Driver> drivers;
    private Map<String, Trip> activeTrips;
    private TripManager tripManager;

    private UberSystem() {
        drivers = new ArrayList<>();
        activeTrips = new ConcurrentHashMap<>();
        tripManager = new TripManager();
    }

    public static synchronized UberSystem getInstance() {
        if(instance == null) instance = new UberSystem();
        return instance;
    }

    public void addDriver(Driver d) { drivers.add(d); }

    public Trip requestRide(Rider rider, Location src, Location dest, RideType type) {
        Driver driver = tripManager.findNearestDriver(src, drivers);
        if(driver == null) {
            throw new RuntimeException("No Drivers Available!");
        }
        
        driver.setStatus(DriverStatus.ON_TRIP);
        
        PricingStrategy strategy = (type == RideType.PREMIUM) ? new PremiumPricing() : new RegularPricing();
        Trip trip = new Trip(UUID.randomUUID().toString(), rider, driver, src, dest, strategy);
        
        activeTrips.put(trip.getId(), trip);
        return trip;
    }

    public void completeTrip(String tripId) {
        Trip trip = activeTrips.get(tripId);
        if(trip != null) {
            trip.setStatus(TripStatus.COMPLETED);
            trip.getDriver().setStatus(DriverStatus.AVAILABLE);
            // In real app: Update Driver Location to Dest
            trip.getDriver().setLocation(trip.dest);
        }
    }
}

File: LLD/RideSharingSystem.py
# fmt: off
# ==============================================================================
#  RIDE SHARING SYSTEM — ASCII CLASS DIAGRAM
# ==============================================================================
#
#  ┌──────────────────────────────────────────────────────────────────────────┐
#  │                    RIDE SHARING SYSTEM (Uber-like)                       │
#  └──────────────────────────────────────────────────────────────────────────┘
#
#  ┌────────────────────────────┐
#  │    RideSharingService      │  ← Facade
#  ├────────────────────────────┤
#  │ + riders: Dict             │
#  │ + drivers: Dict            │
#  │ + trips: Dict              │
#  │ + pricing: PricingStrategy │
#  ├────────────────────────────┤
#  │ + register_rider()         │
#  │ + register_driver()        │
#  │ + request_ride()           │
#  │ + complete_trip()          │
#  │ -_find_nearest_driver()    │
#  └────────────────────────────┘
#            │          │
#        ....│..........│....
#        .   ▼          ▼   .
#  ┌──────────────┐  ┌──────────────────────────┐
#  │    Rider     │  │         Driver           │
#  ├──────────────┤  ├──────────────────────────┤
#  │ + id: str    │  │ + id: str                │
#  │ + name: str  │  │ + name: str              │
#  │ + location   │  │ + location: (x, y)       │
#  └──────────────┘  │ + status: DriverStatus   │
#                    │   (enum)                 │
#                    ├──────────────────────────┤
#                    │ + update_location()      │
#                    └──────────────────────────┘
#
#  ┌────────────────────────────────────────────────────┐
#  │                       Trip                         │
#  ├────────────────────────────────────────────────────┤
#  │ + id: str                                          │
#  │ + rider: Rider                                     │
#  │ + driver: Driver                                   │
#  │ + pickup_loc, dropoff_loc: (x, y)                  │
#  │ + status: TripStatus (enum)                        │
#  │ + fare: float                                      │
#  │ + start_time, end_time: datetime                   │
#  ├────────────────────────────────────────────────────┤
#  │ + start()                                          │
#  │ + complete(pricing_strategy)                       │
#  └────────────────────────────────────────────────────┘
#
#  ┌─────────────────────────────────┐
#  │       PricingStrategy           │  ← Strategy Pattern
#  │         (ABC/Interface)         │
#  ├─────────────────────────────────┤
#  │ + calculate(trip): float        │
#  └─────────────────┬───────────────┘
#                    │
#             ┌──────┴──────┐
#             │             │
#             ▼             ▼
#  ┌───────────────────┐ ┌────────────────────────┐
#  │  RegularPricing   │ │   SurgePricing         │
#  ├───────────────────┤ ├────────────────────────┤
#  │ BASE=2.0 +        │ │ SURGE_MULTIPLIER=2.5   │
#  │ 1.0 per unit dist │ │ applied on top of base │
#  └───────────────────┘ └────────────────────────┘
#
#  ┌─────────────────────────┐  ┌─────────────────────┐
#  │   DriverStatus (Enum)   │  │  TripStatus (Enum)  │
#  ├─────────────────────────┤  ├─────────────────────┤
#  │  AVAILABLE              │  │  REQUESTED          │
#  │  BUSY                   │  │  IN_PROGRESS        │
#  │  OFFLINE                │  │  COMPLETED          │
#  └─────────────────────────┘  │  CANCELLED          │
#                               └─────────────────────┘
#
#  RELATIONSHIPS:
#  RideSharingService ──*──> Rider          (registered riders)
#  RideSharingService ──*──> Driver         (registered drivers)
#  RideSharingService ──*──> Trip           (active and past trips)
#  RideSharingService ──1──> PricingStrategy(swappable pricing)
#  Trip ──1──> Rider                        (who requested)
#  Trip ──1──> Driver                       (who serves)
#  RegularPricing / SurgePricing ──▷── PricingStrategy (implements)
#  Driver.status transitions: AVAILABLE → BUSY (on ride start) → AVAILABLE (on complete)
# ==============================================================================
# fmt: on
import threading
import uuid
import math
from abc import ABC, abstractmethod
from enum import Enum
from typing import List, Dict, Optional
from datetime import datetime

"""
==============================================================================================
RIDE SHARING SERVICE (LIKE UBER) LOW LEVEL DESIGN (INTERVIEW OPTIMIZED)
==============================================================================================

Key Features Implemented:
1. User Management: Rider and Driver (Extends User).
2. Geo-Location: Distance-based calculation and nearest driver matching.
3. Trip Lifecycle: REQUESTED -> ASSIGNED -> ON_TRIP -> COMPLETED.
4. Cost Calculation: Strategy pattern for Regular/Premium pricing.
5. Concurrency: Thread-safe driver status updates.

Design Patterns:
1. Facade: RideSharingService (Central Controller).
2. Strategy: PricingStrategy (Regular/Premium).
3. State: Handled via TripStatus and DriverStatus.

Class Design Diagram:
---------------------
[RideSharingService] "1" *-- "*" [Driver]
[RideSharingService] "1" *-- "*" [Rider]
[RideSharingService] "1" *-- "*" [Trip]
[Trip] "1" *-- "1" [Driver]
[Trip] "1" *-- "1" [Rider]
[Trip] "1" *-- "1" [PricingStrategy]
[Driver] <|-- [User]
[Rider] <|-- [User]
[PricingStrategy] <|-- [RegularPricing]
[PricingStrategy] <|-- [PremiumPricing]

Class Details:
---------------------
1. RideSharingService (Facade)
   - Role: Main controller.
   - Methods: requestRide(), completeRide().

2. Trip
   - Attributes: id, rider, driver, src, dest, status, fare.
   - Methods: startTrip(), completeTrip().

3. Driver
   - Attributes: status (AVAILABLE, BUSY), location.
   - Methods: updateStatus().

4. PricingStrategy
   - Role: Calculate fare based on distance and time.
   - Impls: RegularPricing, PremiumPricing.
"""

# ==========================================
# Enums & Models
# ==========================================

class RideType(Enum):
    REGULAR = "REGULAR"
    PREMIUM = "PREMIUM"

class TripStatus(Enum):
    ASSIGNED = "ASSIGNED"
    ON_TRIP = "ON_TRIP"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"

class DriverStatus(Enum):
    AVAILABLE = "AVAILABLE"
    BUSY = "BUSY"
    OFFLINE = "OFFLINE"

class Location:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def distance_to(self, other: 'Location') -> float:
        """Calculate Euclidean distance between two points."""
        return math.sqrt((self.x - other.x)**2 + (self.y - other.y)**2)

    def __repr__(self):
        return f"({self.x}, {self.y})"

# ==========================================
# Pricing Strategies
# ==========================================

class PricingStrategy(ABC):
    @abstractmethod
    def calculate_fare(self, distance: float, time_mins: float) -> float:
        pass

class RegularPricing(PricingStrategy):
    def calculate_fare(self, distance: float, time_mins: float) -> float:
        return (distance * 10.0) + (time_mins * 2.0)

class PremiumPricing(PricingStrategy):
    def calculate_fare(self, distance: float, time_mins: float) -> float:
        return (distance * 20.0) + (time_mins * 4.0)

# ==========================================
# Domain Models
# ==========================================

class User:
    def __init__(self, user_id: str, name: str):
        self.id = user_id
        self.name = name

class Rider(User):
    def __init__(self, user_id: str, name: str, location: Location):
        super().__init__(user_id, name)
        self.current_location = location

class Driver(User):
    def __init__(self, user_id: str, name: str, location: Location):
        super().__init__(user_id, name)
        self.current_location = location
        self.status = DriverStatus.AVAILABLE
        self.rating = 5.0
        self._lock = threading.Lock()

    def update_status(self, status: DriverStatus):
        with self._lock:
            self.status = status
            print(f"INFO: Driver {self.name} status -> {status.name}")

class Trip:
    """Represents a single ride journey."""
    def __init__(self, rider: Rider, driver: Driver, src: Location, dest: Location, pricing: PricingStrategy):
        self.id = str(uuid.uuid4())
        self.rider = rider
        self.driver = driver
        self.src = src
        self.dest = dest
        self.pricing = pricing
        self.status = TripStatus.ASSIGNED
        self.fare = 0.0

    def start_trip(self):
        self.status = TripStatus.ON_TRIP
        print(f"INFO: Trip {self.id[:8]}... started. {self.rider.name} with {self.driver.name}")

    def complete_trip(self):
        self.status = TripStatus.COMPLETED
        distance = self.src.distance_to(self.dest)
        self.fare = self.pricing.calculate_fare(distance, 15.0)  # 15min assumed for demo
        print(f"INFO: Trip completed. Fare: ${self.fare:.2f}")

# ==========================================
# System Facade
# ==========================================

class RideSharingService:
    def __init__(self):
        self.drivers: List[Driver] = []
        self.riders: Dict[str, Rider] = {}
        self.active_trips: Dict[str, Trip] = {}
        print("INFO: RideSharingService initialized.")

    def add_driver(self, driver: Driver):
        self.drivers.append(driver)

    def add_rider(self, rider: Rider):
        self.riders[rider.id] = rider

    def request_ride(self, rider_id: str, src: Location, dest: Location, ride_type: RideType) -> Optional[Trip]:
        """Finds the nearest driver and creates a ride request."""
        rider = self.riders.get(rider_id)
        if not rider:
            print(f"ERROR: Rider {rider_id} not registered.")
            return None

        best_driver = self._find_nearest_driver(src)
        if not best_driver:
            print("WARNING: No available drivers nearby.")
            return None

        best_driver.update_status(DriverStatus.BUSY)
        pricing = PremiumPricing() if ride_type == RideType.PREMIUM else RegularPricing()
        trip = Trip(rider, best_driver, src, dest, pricing)
        self.active_trips[trip.id] = trip
        print(f"INFO: Ride assigned! Driver {best_driver.name} -> {rider.name}")
        return trip

    def _find_nearest_driver(self, location: Location) -> Optional[Driver]:
        available = [d for d in self.drivers if d.status == DriverStatus.AVAILABLE]
        if not available:
            return None
        return min(available, key=lambda d: d.current_location.distance_to(location))

    def complete_ride(self, trip_id: str):
        trip = self.active_trips.get(trip_id)
        if trip:
            trip.complete_trip()
            trip.driver.update_status(DriverStatus.AVAILABLE)
            trip.driver.current_location = trip.dest
            del self.active_trips[trip_id]
        else:
            print(f"ERROR: Trip {trip_id} not found.")

# ==========================================
# Main Execution / Demo
# ==========================================

if __name__ == "__main__":
    print("--- Starting Ride Sharing System Demo ---")

    service = RideSharingService()

    d1 = Driver("D1", "Bob", Location(1, 1))
    d2 = Driver("D2", "Charlie", Location(10, 10))
    service.add_driver(d1)
    service.add_driver(d2)

    rider1 = Rider("R1", "Alice", Location(0, 0))
    service.add_rider(rider1)

    print("[Action] Alice requests a REGULAR ride to (5, 5)")
    trip = service.request_ride("R1", rider1.current_location, Location(5, 5), RideType.REGULAR)

    if trip:
        trip.start_trip()
        service.complete_ride(trip.id)

    print(f"INFO: Bob status: {d1.status.name} at Location: {d1.current_location}")
